src/connector/src/sink/elasticsearch_opensearch/elasticsearch_opensearch_common.rs

```diff
@@ -0,0 +1,594 @@
+pub const ES_OPTION_ROUTING_COLUMN: &str = "routing_column";
+
+#[serde_as]
+#[derive(Deserialize, Debug, Clone, WithOptions)]
+pub struct ElasticSearchOpenSearchConfig {
+    #[serde(rename = "url")]
+    pub url: String,
+    /// The index's name of elasticsearch or openserach
+    #[serde(rename = "index")]
+    pub index: Option<String>,
+    /// If pk is set, then "pk1+delimiter+pk2+delimiter..." will be used as the key, if pk is not set, we will just use the first column as the key.
+    #[serde(rename = "delimiter")]
+    pub delimiter: Option<String>,
+    /// The username of elasticsearch or openserach
+    #[serde(rename = "username")]
+    pub username: String,
+    /// The username of elasticsearch or openserach
+    #[serde(rename = "password")]
+    pub password: String,
+    /// It is used for dynamic index, if it is be set, the value of this column will be used as the index. It and `index` can only set one
+    #[serde(rename = "index_column")]
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    pub index_column: Option<usize>,
+
+    /// It is used for dynamic route, if it is be set, the value of this column will be used as the route
+    #[serde(rename = "routing_column")]
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    pub routing_column: Option<usize>,
+
+    #[serde(rename = "retry_on_conflict")]
+    #[serde_as(as = "DisplayFromStr")]
+    #[serde(default = "default_retry_on_conflict")]
+    pub retry_on_conflict: i32,
+
+    #[serde(rename = "batch_num_messages")]
+    #[serde_as(as = "DisplayFromStr")]
+    #[serde(default = "default_batch_num_messages")]
+    pub batch_num_messages: usize,
+
+    #[serde(rename = "batch_size_kb")]
+    #[serde_as(as = "DisplayFromStr")]
+    #[serde(default = "default_batch_size_kb")]
+    pub batch_size_kb: usize,
+
+    #[serde(rename = "concurrent_requests")]
+    #[serde_as(as = "DisplayFromStr")]
+    #[serde(default = "default_concurrent_requests")]
+    pub concurrent_requests: usize,
+}
+
+fn default_retry_on_conflict() -> i32 {
+    3
+}
+
+fn default_batch_num_messages() -> usize {
+    512
+}
+
+fn default_batch_size_kb() -> usize {
+    5 * 1024
+}
+
+fn default_concurrent_requests() -> usize {
+    1024
+}
+
+impl ElasticSearchOpenSearchConfig {
+    pub fn from_btreemap(
+        mut properties: BTreeMap<String, String>,
+        schema: &Schema,
+    ) -> Result<Self> {
+        Self::covert_name_to_index(&mut properties, ES_OPTION_INDEX_COLUMN, schema)?;
+        Self::covert_name_to_index(&mut properties, ES_OPTION_ROUTING_COLUMN, schema)?;
+        let config = serde_json::from_value::<ElasticSearchOpenSearchConfig>(
+            serde_json::to_value(properties).unwrap(),
+        )
+        .map_err(|e| SinkError::Config(anyhow!(e)))?;
+        Ok(config)
+    }
+
+    fn covert_name_to_index(
+        properties: &mut BTreeMap<String, String>,
+        key_name: &str,
+        schema: &Schema,
+    ) -> Result<()> {
+        let index_column = properties
```

Comment:
The user input is option<string>, which I changed in the from_btreemap, similar to the primary_key operation.

Reply:
I know. What I meant was that, the type of field `index_column` and `routing_column` in `ElasticSearchOpenSearchConfig` should be `Option<String>`, because the `ElasticSearchOpenSearchConfig` should be consistent to the user facing config, and it's wired to manually modify the config here.